File: examples/Delete/Callback.cpp
#include <map>
#include <iostream>
#include <omero/client.h>
#include <omero/callbacks.h>
#include <omero/cmd/Graphs.h>

using namespace std;

namespace OA = omero::api;
namespace OCB = omero::callbacks;
namespace OCMD = omero::cmd;


/**
 * Uses the default {@link DeleteCallbackI} instance.
 */
int main(int argc, char* argv[]) {

    omero::client_ptr c = new omero::client(); // Close handled by destructor
    OA::ServiceFactoryPrx s = c->createSession();

    {
        OCMD::DeletePtr dc = new OCMD::Delete();
        dc->type = "/Image";
        dc->id = 1;

        OCMD::HandlePrx deleteHandlePrx = s->submit(dc);
        OCB::CmdCallbackIPtr cb = new OCB::CmdCallbackI(c, deleteHandlePrx, true);  // Closed by destructor

        try {
            cb->loop(10, 500);

            OCMD::DeleteRspPtr rsp = OCMD::DeleteRspPtr::dynamicCast(
                    cb->getResponse());

            if (rsp) {
                // Not an error state then.
                cout << "Report:  deleted=" << rsp->actualDeletes;
                if (!rsp->warning.empty()) {
                    cout << ", warning=" << rsp->warning;
                }
            }
        } catch (const omero::LockTimeout& lt) {
            cout << "Not finished in 5 seconds. Cancelling..." << endl;
            if (!deleteHandlePrx->cancel()) {
                cout << "ERROR: Failed to cancel" << endl;
            }
        }
    }
}

File: examples/Delete/Callback.java
import java.util.List;

import omero.LockTimeout;
import omero.ServerError;
import omero.api.ServiceFactoryPrx;
import omero.cmd.CmdCallbackI;
import omero.cmd.Delete2;
import omero.cmd.HandlePrx;
import Glacier2.CannotCreateSessionException;
import Glacier2.PermissionDeniedException;

import com.google.common.collect.ImmutableListMultimap;

/**
 * Uses the default {@link DeleteCallbackI} instance.
 */
public class Callback {

    public static void main(String[] args) throws CannotCreateSessionException,
            PermissionDeniedException, ServerError {

        omero.client c = new omero.client();
        ServiceFactoryPrx s = c.createSession();

        try {
            Map<String, List<Long>> 
            Delete2 dc = new Delete2(
                    new ImmutableListMultimap.Builder<String, Long>()
                        .put("Image", 1l).build().asMap(), null, false /* non-dry-run */);
            // TODO: refactor to c.submitAndWait()
            HandlePrx deleteHandlePrx = c.getSession().submit(dc);
            CmdCallbackI cb = new CmdCallbackI(c, deleteHandlePrx);
            try {

                cb.loop(10, 500);

                DeleteReport[] reports = deleteHandlePrx.report();
                DeleteReport r = reports[0]; // We only sent one command
                System.out.println(String.format(
                        "Report:error=%s,warning=%s,deleted=%s", r.error,
                        r.warning, r.actualDeletes));

            } catch (LockTimeout lt) {
                System.out.println("Not finished in 5 seconds. Cancelling...");
                if (!deleteHandlePrx.cancel()) {
                    System.out.println("ERROR: Failed to cancel");
                }
            } finally {
                cb.close(True); // close handle
            }

        } finally {
            c.closeSession();
        }

    }

}
